olp-cpp-sdk-core/src/network2/winhttp/NetworkProtocolWinHttp.cpp

```diff
@@ -0,0 +1,1063 @@
+    case ERROR_WINHTTP_LOGIN_FAILURE:
+      return "Login failed";
+    case ERROR_WINHTTP_NAME_NOT_RESOLVED:
+      return "Name not resolved";
+    case ERROR_WINHTTP_OPERATION_CANCELLED:
+      return "Cancelled";
+    case ERROR_WINHTTP_REDIRECT_FAILED:
+      return "Redirect failed";
+    case ERROR_WINHTTP_RESEND_REQUEST:
+      return "Resend request";
+    case ERROR_WINHTTP_RESPONSE_DRAIN_OVERFLOW:
+      return "Response overflow";
+    case ERROR_WINHTTP_SECURE_FAILURE:
+      return "Security error";
+    case ERROR_WINHTTP_TIMEOUT:
+      return "Timed out";
+    case ERROR_WINHTTP_UNRECOGNIZED_SCHEME:
+      return "Invalid scheme";
+  }
+  return "Unknown error";
+}
+
+olp::network2::ErrorCode errorToCode(DWORD err) {
+  if (err == ERROR_SUCCESS)
+    return olp::network2::ErrorCode::SUCCESS;
+  else if ((err == ERROR_WINHTTP_INVALID_URL) ||
+           (err == ERROR_WINHTTP_UNRECOGNIZED_SCHEME) ||
+           (err == ERROR_WINHTTP_NAME_NOT_RESOLVED))
+    return olp::network2::ErrorCode::INVALID_URL_ERROR;
+  else if ((err == ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED) ||
+           (err == ERROR_WINHTTP_LOGIN_FAILURE) ||
+           (err == ERROR_WINHTTP_SECURE_FAILURE))
+    return olp::network2::ErrorCode::AUTHORIZATION_ERROR;
+  else if (err == ERROR_WINHTTP_OPERATION_CANCELLED)
+    return olp::network2::ErrorCode::CANCELLED_ERROR;
+  else if (err == ERROR_WINHTTP_TIMEOUT)
+    return olp::network2::ErrorCode::TIMEOUT_ERROR;
+  return olp::network2::ErrorCode::IO_ERROR;
+}
+
+LPWSTR
+queryHeadervalue(HINTERNET request, DWORD header) {
+  DWORD len = 0, index = WINHTTP_NO_HEADER_INDEX;
+  if (WinHttpQueryHeaders(request, header, WINHTTP_HEADER_NAME_BY_INDEX,
+                          WINHTTP_NO_OUTPUT_BUFFER, &len, &index))
+    return NULL;
+  if (GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
+    return NULL;
+  }
+  LPWSTR buffer = (LPWSTR)LocalAlloc(LPTR, len);
+  if (!buffer) return NULL;
+  index = WINHTTP_NO_HEADER_INDEX;
+  if (!WinHttpQueryHeaders(request, header, WINHTTP_HEADER_NAME_BY_INDEX,
+                           buffer, &len, &index)) {
+    LocalFree(buffer);
+    return NULL;
+  }
+  return buffer;
+}
+
+void UnixTimeToFileTime(std::uint64_t t, LPFILETIME pft) {
+  // Note that LONGLONG is a 64-bit value
+  LONGLONG ll;
+
+  ll = Int32x32To64(t, 10000000) + 116444736000000000;
+  pft->dwLowDateTime = (DWORD)ll;
+  pft->dwHighDateTime = ll >> 32;
+}
+
+bool convertMultiByteToWideChar(const std::string& s_in, std::wstring& s_out) {
+  s_out.clear();
+  if (s_in.empty()) {
+    return true;
+  }
+
+  // Detect required buffer size.
+  const auto chars_required = MultiByteToWideChar(
+      CP_ACP, MB_PRECOMPOSED, s_in.c_str(),
+      -1,       // denotes null-terminated string
+      nullptr,  // output buffer is null means request required buffer size
+      0);
+
+  if (chars_required == 0) {
+    // error: could not convert input string
+    return false;
+  }
+
+  s_out.resize(chars_required);
```

Comment:
Maybe create a temporary `std::wstring`, use it in `MultiByteToWideChar ` and then swap it with `s_out` (or just `std::move`-it)? 

```suggestion
std::wstring temp;
temp.resize(chars_required);
// perform MultiByteToWideChar
if (conversion_result == 0) {
    return false;
}
s_out = std::move(temp);
return true;
```

So, it doesn't have any side-effect on `s_out`, in case conversion failed?